design/cluster-scope-resource-filter.md

```diff
@@ -0,0 +1,410 @@
+	--include-namespaced-resources=persistentvolumeclaim,deployment,service,endpoint,pod,replicaset
+```
+
+Example of old parameters:
+``` bash
+velero backup create <backup-name>
+	--include-resources=persistentvolumeclaim,deployment,service,endpoint,pod,replicaset
+```
+
+The following commands mean backup PVC, Deployment, Service, Endpoint, Pod and ReplicaSet resources in namespaces default and kube-system, and related cluster resources. PVC related PV is included too.
+
+Example of new parameters:
+``` bash
+velero backup create <backup-name>
+	--include-namespaces=default,kube-system
+	--include-namespaced-resources=persistentvolumeclaim,deployment,service,endpoint,pod,replicaset
+```
+
+Example of old parameters:
+``` bash
+velero backup create <backup-name>
+	--include-namespaces=default,kube-system
+	--include-resources=persistentvolumeclaim,deployment,service,endpoint,pod,replicaset
+```
+
+This means backup all resources except Ingress type resources in all namespaces, and related cluster resources.
+
+Example of new parameters:
+``` bash
+velero backup create <backup-name>
+	--exclude-namespaced-resources=ingress
+```
+
+Example of old parameters:
+``` bash
+velero backup create <backup-name>
+	--exclude-resources=ingress
+```
+
+#### some namespaced resources + some additional cluster resources
+This means backup all resources in namespace in default, kube-system, and related cluster resources, plus all StorageClass cluster resources.
+``` bash
+velero backup create <backup-name>
+	--include-namespaces=default,kube-system
+	--include-cluster-scope-resources=storageclass
+```
+
+This means backup PVC, Deployment, Service, Endpoint, Pod and ReplicaSet resources in all namespaces, and related cluster resources, plus all StorageClass cluster resources, and PVC related PV.
+``` bash
+velero backup create <backup-name>
+	--include-namespaced-resources=persistentvolumeclaim,deployment,service,endpoint,pod,replicaset
+	--include-cluster-scope-resources=storageclass
+```
+
+This means backup PVC, Deployment, Service, Endpoint, Pod and ReplicaSet resources in default and kube-system namespaces, and related cluster resources, plus all StorageClass cluster resources, and PVC related PV.
+``` bash
+velero backup create <backup-name>
+	--include-namespaced-resources=persistentvolumeclaim,deployment,service,endpoint,pod,replicaset
+	--include-namespaces=default,kube-system
+	--include-cluster-scope-resources=storageclass
+```
+
+This means backup PVC, Deployment, Service, Endpoint, Pod and ReplicaSet resources in default and kube-system namespaces, and related cluster resources, plus all cluster scope resources except StorageClass type resources.
+``` bash
+velero backup create <backup-name>
+	--include-namespaced-resources=persistentvolumeclaim,deployment,service,endpoint,pod,replicaset
+	--include-namespaces=default,kube-system
+	--exclude-cluster-scope-resources=storageclass
+```
+
+#### some namespaced resources + all cluster resources
```

Comment:
@ihcsim
Thanks for your input.
I agree that these new parameters would make Velero CLI more complicated.

I'm not sure about how to include all cluster scope resources except several specific resources and have all namespace resources in a namespace with your proposal.
I saw your feeling about this case. Your suggestion is separate into different backups, but there are also some user requirements for this using scenario.

Reply:
I think it's not just the CLI, presumably, the API will be complicated too, right? Do we know what the APIs will look like?

>  include all cluster scope resources except several specific resources and have all namespace resources in a namespace with your proposal.

Is there a limitation on `--exclude-resources` that prohibits us from achieving this with something like:
```sh
velero backup create <backup-name> \
    --include-cluster-resources=true \
    --include-namespaces=default,kube-system \
    --exclude-resources=persistentvolumeclaim,persistentvolume,customresourcedefinitions
```

Reply:
Yes. There would be more parameters. Please check the detailed design part. There is an example of `BackupSpec` change. But they are separated into two groups, and they cannot work together. They are supposed to be replacements for old parameters. Although more parameters make the code logic more complicated, that would not be a burden for the users to understand.

@sseago @reasonerjt 
What's your opinion about Ivan's proposal? I didn't go through all the scenarios yet. I think even not working for all possibilities by now, after some modification, it should do the same job, but in a quite different way.
I think that depends on the purpose. If we want to enhance the current parameters and keep existing behavior, Ivan's proposal is better. If we want to re-implement the filters, I think the current proposal is better.

Reply:
What are the reasons behind replacing/deprecating the old flags? Feels like it will be quite a process since they are commonly used. Following the ongoing discussion, deciding on the default behavior of the new flags feels complicated enough already. Feels like this will impose quite a bit of cognitive burden on the users.

Reply:
@ihcsim See the linked issue for the discussion around this. Short answer is that there are several desired use cases that are impossible to specify with the old set of parameters, and the new parameters provide the flexibility to meet them: https://github.com/vmware-tanzu/velero/issues/5120

Reply:
This is the original issue to request this feature https://github.com/vmware-tanzu/velero/issues/5120.
To support filtering both the cluster-scoped resources and namespace resources, we need to do some modifications to the filters.
To me, either way, it will raise the using complexity. Just need to choose which way is better or preferred.

Reply:
@blackpiglet As for the proposal overall, I think it's at least as complicated as what we're proposing here. To see whether it covers all of the use cases, someone would need to go through all of the use cases that we're solving in this proposal and see whether there are any gaps there. Overall, I think this proposal feels more intuitive -- it's less of a radical departure from current behavior. As for proliferation of configuration options, it's important to note that a user will either use the old params or the new ones, not both -- so instead of at most 3 params in use, there will be at most 4 now.

Reply:
The benefit of introducing new options to augment the behaviour of existing include/exclude/include-cluster resources options is that users don't have to re-learn a new filtering system, nor modify their existing configuration if they choose not to.

On the other hand, I do think the existing filtering options are limited for nowadays' requirements. If we are going to invest the time into enhancing this, we should do it with the hope of not having to re-do this exercise again in say, 2 years time.

If we are going to deprecate **and then remove** the older options (especially, `--include-cluster-resources`), the community should be given sufficient time to adapt the changes. E.g., 2 releases with deprecation warning, then removal from CLI, then finally removal from API. We should also think through the upgrade path.

Reply:
@ihcsim 
FYI, if this PR is merged, we can still trace the topic you mentioned in issue #5120.